packages/blocks/src/api/test/registration.js

```diff
@@ -304,6 +304,56 @@ describe( 'blocks', () => {
 			} );
 		} );
 
+		it( 'should default to empty object when attributes is omitted and without warning', () => {
+			registerBlockType( 'core/test-block-omitted-attributes', {
+				title: 'block title',
+				category: 'text',
+				save: noop,
+			} );
+
+			// Verify no warning was shown (unlike when explicitly set to null/undefined)
+			expect( console ).not.toHaveWarned();
+
+			const blockType = getBlockType(
+				'core/test-block-omitted-attributes'
+			);
+			expect( blockType.attributes ).toEqual( {} );
+		} );
+
+		it( 'should warn and default to empty object when attributes is undefined', () => {
+			registerBlockType( 'core/test-block-undefined-attributes', {
+				title: 'block title',
+				category: 'text',
+				save: noop,
+				attributes: undefined,
+			} );
+
+			expect( console ).toHaveWarnedWith(
+				'The block "core/test-block-undefined-attributes" must not register attributes as `null` or `undefined`. Use an empty object (`attributes: {}`) or exclude the `attributes` key. In the next Gutenberg release, passing `null` or `undefined` for `attributes` will cause the block editor to crash.'
+			);
+
+			const blockType = getBlockType(
+				'core/test-block-undefined-attributes'
+			);
+			expect( blockType.attributes ).toEqual( {} );
+		} );
+
+		it( 'should warn and default to empty object when attributes is null', () => {
+			registerBlockType( 'core/test-block-null-attributes', {
+				title: 'block title',
+				category: 'text',
+				save: noop,
+				attributes: null,
+			} );
+
+			expect( console ).toHaveWarnedWith(
+				'The block "core/test-block-null-attributes" must not register attributes as `null` or `undefined`. Use an empty object (`attributes: {}`) or exclude the `attributes` key. In the next Gutenberg release, passing `null` or `undefined` for `attributes` will cause the block editor to crash.'
+			);
+
+			const blockType = getBlockType( 'core/test-block-null-attributes' );
+			expect( blockType.attributes ).toEqual( {} );
+		} );
```

Comment:
We could use `it.each()` and remove some duplication:



```suggestion
		it.each( [
			[ 'undefined', undefined ],
			[ 'null', null ],
		] )(
			'should warn and default to empty object when attributes is %s',
			( _label, value ) => {
				registerBlockType( 'core/test-block-null-attributes', {
					title: 'block title',
					category: 'text',
					save: noop,
					attributes: value,
				} );

				expect( console ).toHaveWarnedWith(
					'The block "core/test-block-null-attributes" must not register attributes as `null` or `undefined`. Use an empty object (`attributes: {}`) or exclude the `attributes` key. In the next Gutenberg release, passing `null` or `undefined` for `attributes` will cause the block editor to crash.'
				);

				const blockType = getBlockType(
					'core/test-block-null-attributes'
				);
				expect( blockType.attributes ).toEqual( {} );
			}
		);
```